generated/nidaqmx/nidaqmx_service.cpp

```diff
@@ -14001,15 +14001,15 @@ namespace nidaqmx_grpc {
     using CallbackRouter = nidevice_grpc::CallbackRouter<int32, TaskHandle, int32>;
     class RegisterDoneEventReactor : public nidevice_grpc::ServerWriterReactor<RegisterDoneEventResponse, nidevice_grpc::CallbackRegistration> {
     public:
-    RegisterDoneEventReactor(const RegisterDoneEventRequest& request, NiDAQmxLibraryInterface* library, const ResourceRepositorySharedPtr& session_repository)
+    RegisterDoneEventReactor(::grpc::CallbackServerContext* context, const RegisterDoneEventRequest* request, NiDAQmxLibraryInterface* library, NiDAQmxService* service)
     {
-      auto status = start(&request, library, session_repository);
+      auto status = start(context, request, library, service);
       if (!status.ok()) {
         this->Finish(status);
       }
```

Comment:
This asserted while testing with nidaqmx-python:
![image](https://github.com/ni/grpc-device/assets/17914399/eab083a8-6100-4ae7-bdd1-c1d18e0f99c3)

```
 	win32u.dll!NtUserWaitMessage()	Unknown
 	user32.dll!DialogBox2()	Unknown
 	user32.dll!InternalDialogBox()	Unknown
 	user32.dll!SoftModalMessageBox()	Unknown
 	user32.dll!MessageBoxWorker(struct _MSGBOXDATA *)	Unknown
 	user32.dll!MessageBoxTimeoutW()	Unknown
 	user32.dll!MessageBoxW()	Unknown
 	ucrtbased.dll!__acrt_MessageBoxW(HWND__ * hwnd, const wchar_t * text, const wchar_t * caption, unsigned int type) Line 695	C++
 	ucrtbased.dll!__crt_char_traits<wchar_t>::message_box<HWND__ * __ptr64,wchar_t const * __ptr64 const & __ptr64,wchar_t const * __ptr64 const & __ptr64,unsigned int const & __ptr64>(HWND__ * && <args_0>, const wchar_t * const & <args_1>, const wchar_t * const & <args_2>, const unsigned int & <args_3>) Line 121	C++
 	ucrtbased.dll!common_show_message_box<wchar_t>(const wchar_t * const text, const wchar_t * const caption, const unsigned int type) Line 117	C++
 	ucrtbased.dll!__acrt_show_wide_message_box(const wchar_t * text, const wchar_t * caption, unsigned int type) Line 135	C++
 	ucrtbased.dll!common_message_window<wchar_t>(const int report_type, void * const return_address, const wchar_t * const file_name, const wchar_t * const line_number, const wchar_t * const module_name, const wchar_t * const user_message) Line 404	C++
 	ucrtbased.dll!__acrt_MessageWindowW(const int report_type, void * const return_address, const wchar_t * const file_name, const wchar_t * const line_number, const wchar_t * const module_name, const wchar_t * const user_message) Line 452	C++
 	ucrtbased.dll!_VCrtDbgReportW(int nRptType, void * returnAddress, const wchar_t * szFile, int nLine, const wchar_t * szModule, const wchar_t * szFormat, char * arglist) Line 642	C++
 	ucrtbased.dll!_CrtDbgReportW(int report_type, const wchar_t * file_name, int line_number, const wchar_t * module_name, const wchar_t * format, ...) Line 273	C++
 	ucrtbased.dll!issue_debug_notification(const wchar_t * const message) Line 122	C++
 	ucrtbased.dll!__acrt_report_runtime_error(const wchar_t * message) Line 142	C++
 	ucrtbased.dll!abort() Line 61	C++
 	ni_grpc_device_server.exe!grpc::CoreCodegen::assert_fail(const char * failed_assertion, const char * file, int line) Line 246	C++
 	ni_grpc_device_server.exe!grpc::internal::CallbackWithSuccessTag::Set(grpc_call * call, std::function<void __cdecl(bool)> f, grpc::internal::CompletionQueueTag * ops, bool can_inline) Line 164	C++
>	ni_grpc_device_server.exe!grpc::internal::CallbackServerStreamingHandler<nidaqmx_grpc::RegisterDoneEventRequest,nidaqmx_grpc::RegisterDoneEventResponse>::ServerCallbackWriterImpl::Finish(grpc::Status s) Line 523	C++
 	ni_grpc_device_server.exe!grpc::ServerWriteReactor<nidaqmx_grpc::RegisterDoneEventResponse>::Finish(grpc::Status s) Line 648	C++
 	ni_grpc_device_server.exe!nidevice_grpc::ServerWriterReactor<nidaqmx_grpc::RegisterDoneEventResponse,nidevice_grpc::CallbackRegistration>::OnCancel() Line 43	C++
 	ni_grpc_device_server.exe!<lambda_2f27ce0846995826ab9faa5139791063>::<lambda_invoker_cdecl>(void * void_arg, absl::lts_20220623::Status __formal) Line 70	C++
 	ni_grpc_device_server.exe!grpc_core::Executor::RunClosures(const char * executor_name, grpc_closure_list list) Line 127	C++
 	ni_grpc_device_server.exe!grpc_core::Executor::ThreadMain(void * arg) Line 245	C++
 	ni_grpc_device_server.exe!`anonymous namespace'::ThreadInternalsWindows::thread_body(void * v) Line 122	C++
 	kernel32.dll!BaseThreadInitThunk()	Unknown
 	ntdll.dll!RtlUserThreadStart()	Unknown
```

I think the problem is that we are calling `Finish()` twice:
- In the `RegisterXEventReactor` constructor because `start()` returned a not-ok status. 
- Again in `ServerWriteReactor<TResponse>::OnCancel()`

Calling `Finish()` sets `ServerCallbackWriterImpl::finish_tag_`, which can only be done once.

`grpcpp/support/server_callback.h` doesn't say to call `Finish()` "exactly once", but [grpcpp/support/async_stream.h](https://github.com/grpc/grpc/blob/0526a51734003180862331e2527503a5d1898c95/include/grpcpp/support/async_stream.h#L51) and [grpcpp/support/sync_stream.h](https://github.com/grpc/grpc/blob/0526a51734003180862331e2527503a5d1898c95/include/grpcpp/support/sync_stream.h#L39) do.

Reply:
I wonder why the system tests didn't catch this.

Reply:
Fixed by adding `bool ServerWriteReactor::try_finish(::grpc::Status)`, which only calls `Finish()` once.